File: back/src/main/java/com/openclassrooms/mddapi/service/interfaces/PostService.java
package com.openclassrooms.mddapi.service.interfaces;

import com.openclassrooms.mddapi.dto.requests.CreatePostRequest;

import java.util.List;

import com.openclassrooms.mddapi.dto.PostDto;

public interface PostService {
	
	/**
     * Creates a new post under a specific topic by a logged-in user.
     *
     * @param createPostRequest DTO containing the details of the new post
     * @return PostDto containing the data of the newly created post
     * @throws NotFoundException if the user or topic does not exist
     */
    PostDto createPost(CreatePostRequest createPostRequest);
    
    /**
     * Retrieves a post by its ID, including comments with usernames mapped.
     *
     * @param id the ID of the post to retrieve
     * @return PostDto with details of the post and its comments
     * @throws NotFoundException if the post does not exist
     */
    PostDto getPostById(Long id);
    
    /**
     * Retrieves all posts from topics to which the user is subscribed.
     *
     * @return List of PostDto for all posts in subscribed topics
     * @throws NotFoundException if the user does not exist
     */
    List<PostDto> getPostsBySubscribedTopics();
}

File: front/src/app/core/services/post.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Post } from '../interfaces/posts/post.interface';
import { environment } from 'src/environments/environment';

/**
 * Service responsible for handling API calls related to posts.
 * This service allows fetching, creating, and retrieving specific posts.
 *
 * @Injectable Decorator that marks this class as available for dependency injection,
 * with a singleton instance provided in the root injector.
 */
@Injectable({
  providedIn: 'root'
})
export class PostService {
  private apiUrl = `${environment.baseUrl}/auth/posts`;

  constructor(private http: HttpClient) { }
 /**
   * Fetches a list of posts based on the topics the user is subscribed to.
   * @returns An Observable emitting an array of Post objects.
   */
  getPostsBySubscribedTopics(): Observable<Post[]> {
    return this.http.get<Post[]>(`${this.apiUrl}/feed`);
  }

  /**
   * Retrieves a specific post by its ID.
   * @param postId The unique identifier of the post.
   * @returns An Observable emitting the Post object retrieved.
   */
  getPostById(postId: string): Observable<Post> {
    return this.http.get<Post>(`${this.apiUrl}/${postId}`);
  }
  
   /**
   * Submits a new post to the server.
   * @param post The Post object to be created.
   * @returns An Observable emitting the newly created Post object.
   */
  createPost(post: Post): Observable<Post> {
    return this.http.post<Post>(`${this.apiUrl}/add`, post);
  }
}
